### ISSUE TITLE
empty output when executing `vim.system` in parallel

### ISSUE THREAD

**Issue by vdrn** [neovim/neovim#30846]
### Problem

When I try to run 2 `vim.system` requests at the same time, sometimes the one of the requests returns empty output. 
I'm reproducing the issue when running `nvim --clean` and sourcing the code below.
Using `vim.fn.jobstart` does not produce the same issue.

``` lua
-- test.lua
local nvim_repo = "/home/vedran/source/neovim"

function SystemTest()
  local tries = 10
  for i = 1, tries do
    vim.defer_fn(function()
      local file1 = nil
      local file2 = nil
      vim.system(
        {
          "git",
          "-C", nvim_repo, "cat-file", "-p",
          "b7e896671500a6e9a0c773bc6bac2d073e588eba:src/nvim/tui/tui.c"
        },
        { text = true },
        function(o)
          assert(o.code == 0)
          file1 = o
          if file2 then
            -- print("2nd file finished first. Iter: ", i)
            print("file1", #file1.stdout)
            print("file2", #file2.stdout)
          end
        end
      )
      vim.system(
        { "git", "-C", nvim_repo, "cat-file", "-p", "4846bf05dc:src/nvim/tui/tui.c" },
        { text = true },
        function(o)
          assert(o.code == 0)
          file2 = o
          if file1 then
            -- print("1st file finished first. Iter: ", i)
            print("file1", #file1.stdout)
            print("file2", #file2.stdout)
          end
        end
      )
    end, 0)
  end
end
SystemTest()

```

Example output: 
```
file1 91246
file2 92332

file1 0
file2 92332

file1 91246
file2 92332

file1 91246
file2 92332

file1 91246
file2 92332

file1 91246
file2 0

file1 91246
file2 92332

file1 0
file2 92332

file1 91246
file2 92332

file1 91246
file2 92332
```

Neovim verison:
```
NVIM v0.11.0-dev-988+gfa6ab0d909
Build type: Release
LuaJIT 2.1.1727870382
```

### Steps to reproduce

make sure to set `nvim_repo` in test.lua
`nvim --clean test.lua`
`:source`

### Expected behavior

I expect output to be 
```
file1 91246
file2 92332
```
repeated 10 times

### Nvim version (nvim -v)

v0.11.0-dev-988+gfa6ab0d909

### Vim (not Nvim) behaves the same?

issue is with nvim only api

### Operating system/version

ubuntu 20.04

### Terminal name/version

alacritty 0.12.0-dev (d5b2dac7)

### $TERM environment variable

screen-256color

### Installation

 build from repo

**Comment by lewis6991** [comment]
Are you able to reproduce with `vim.uv.spawn()`?

**Comment by vdrn** [comment]
Yes, with both `vim.uv.spawn()` and `vim.loop.spawn()`.

**Comment by lewis6991** [comment]
This sounds like a uv problem then.

**Comment by vdrn** [comment]
Ok, so I said I can reproduce it with `uv.spawn` because I'm using `plenary.job` to run jobs, and with it I get the same wrong results.
Plenary uses `loop.spawn`, and changing it to `uv.spawn` does not make any difference.

After you asked to update the repro, I've tried implementing the above example with `uv.spawn`, but the result is now correct.

So, this example now produces correct result:
```lua
local nvim_repo = "/home/vedran/source/neovim"

function UvSpawnTest()
  local uv = vim.uv
  local tries = 20
  for i = 1, tries do
    vim.defer_fn(function()
      local file1 = ""
      local file1_done = false
      local file2 = ""
      local file2_done = false

      local stdout1 = uv.new_pipe()
      uv.spawn("git", {
        stdio = { nil, stdout1, nil },
        args = {
          "-C", nvim_repo, "cat-file", "-p",
          "f0973d4227:runtime/lua/vim/lsp/util.lua"
        }
      })
      uv.read_start(stdout1, function(err, data)
        assert(not err, err)
        if data then
          file1 = file1 .. data
        else
          file1_done = true
          if file2_done then
            print("file1", #file1)
            print("file2", #file2)
          end
        end
      end)

      local stdout2 = uv.new_pipe()
      uv.spawn("git",
        {
          stdio = { nil, stdout2, nil },
          args =
          { "-C", nvim_repo, "cat-file", "-p",
            "5bec7288a5b56f1ac38673e59e32f6be774558f3:runtime/lua/vim/lsp/util.lua"
          },
        }
      )
      uv.read_start(stdout2, function(err, data)
        assert(not err, err)
        if data then
          file2 = file2 .. data
        else
          file2_done = true
          if file1_done then
            print("file1", #file1)
            print("file2", #file2)
          end
        end
      end)
    end, 20)
  end
end
UvSpawnTest()
```

And this example using plenary still produces bad result:
```lua
local nvim_repo = "/home/vedran/source/neovim"
function PlenaryTest()
  for name, url in pairs {
    'https://github.com/nvim-lua/plenary.nvim'
  } do
    local install_path = vim.fn.fnamemodify('nvim_issue/' .. name, ':p')
    if vim.fn.isdirectory(install_path) == 0 then
      vim.fn.system { 'git', 'clone', '--depth=1', url, install_path }
    end
    vim.opt.runtimepath:append(install_path)
  end

  local Job = require("plenary.job")
  local tries = 20
  for i = 1, tries do
    vim.defer_fn(function()
      local file1 = nil
      local file2 = nil
      Job:new({
        command = "git",
        args = {
          "-C", nvim_repo, "cat-file", "-p",
          "b7e896671500a6e9a0c773bc6bac2d073e588eba:src/nvim/tui/tui.c"
        },
        on_exit = function(j, c)
          assert(c == 0)
          file1 = j:result()
          if file2 then
            print("file1", #file1)
            print("file2", #file2)
          end
        end,
      }):start()
      Job:new({
        command = "git",
        args = { "-C", nvim_repo, "cat-file", "-p", "4846bf05dc:src/nvim/tui/tui.c" },
        on_exit = function(j, c)
          assert(c == 0)
          file2 = j:result()
          if file1 then
            print("file1", #file1)
            print("file2", #file2)
          end
        end,
      }):start()
    end, 20)
  end
end
PlenaryTest()
```

I'm pretty confused as to whats going on here.
I'll try to figure out the difference between `plenary.job` and my `uv.spawn` example.

I'm not sure do you want me to update the repro steps with plenary example, or keep it as is for now.

**Comment by clason** [comment]
Vim.uv and vim.loop are literally the same thing (just different names for backwards compatibility), namely the main libuv event loop. 

And for the love of all things holy, please let's keep plenary out of this. If it's a Neovim issue, it needs to be reproducible with standard library functions only (otherwise please report at plenary).

**Comment by vdrn** [comment]
I just want to reiterate for posterity:
- `vim.system` is broken
- There is a reproduction in my original post
- Please ignore rest of my comments

**Comment by justinmk** [comment]
Instead of using `print()` in the callbacks to verify the output, can you try collecting all output in a variable. This will avoid any questions about redraw behavior.

**Comment by lewis6991** [comment]
It's still to be determined if `vim.system` is broken or `vim.uv.spawn` is broken.

**Comment by justinmk** [comment]
didn't try. busy.

**Comment by lewis6991** [comment]
Hmm, I wonder whether because it is because `handle:close()` is **only** asynchronous.

At the moment we do:

```lua
---@param handle uv.uv_handle_t?
local function close_handle(handle)
  if handle and not handle:is_closing() then
    handle:close()
  end
end

---@param state vim.SystemState
local function close_handles(state)
  close_handle(state.handle)
  close_handle(state.stdin)
  close_handle(state.stdout)
  close_handle(state.stderr)
  close_handle(state.timer)
end
```

but the docs say

> Synchronous vs Asynchronous Functions ~
>
> Functions that accept a callback are asynchronous. These functions may
> immediately return results to the caller to indicate their initial status, but
> their final execution is deferred until at least the next libuv loop
> iteration. After completion, their callbacks are executed with any results
> passed to it.
>
> Functions that do not accept a callback are synchronous. These functions
> immediately return their results to the caller.
>
> Some (generally FS and DNS) functions can behave either synchronously or
> asynchronously. If a callback is provided to these functions, they behave
> asynchronously; if no callback is provided, they behave synchronously.

Which suggest calling `handle:close()` without a callback is **NOT** synchronous.

So maybe waiting for `state.handle` to fully close before closing the pipes may fix the issue?

**Comment by lewis6991** [comment]
My theory didn't pan out, but closing stdin/stdout fixes the issue I can reproduce.

**Comment by vdrn** [comment]
@lewis6991 
Yeah, https://github.com/neovim/neovim/pull/31442 fixes the issue for me.
